### PR TITLE
feat: add auto-generated Python data structures from ModelPack schema

### LINKED ISSUE #138: Is is possible to auto generate Python APIs?
We have several related discussions here: #48 . What do you think? @caozhuozi

### PR DESCRIPTION
**Pull request by rishi-jat**
## Summary

This PR adds auto-generated Python data structures for the ModelPack specification so downstream Python projects can import the spec types directly instead of copying them manually.

Example usage:

```
python
from model_spec.v1 import Model

model = Model.model_validate_json(json_payload)
print(model.descriptor.name)
```

### Implementation
Python models are generated from the canonical schema: `schema/config-schema.json`

Generation uses datamodel-code-generator and produces Pydantic models.

Added files:
- py/model_spec/v1/models.py – generated Python models
- py/model_spec/v1/__init__.py – exposes the public API
- tools/generate_python_models.py – script to regenerate models
- py/README.md – documentation for usage and regeneration

A Makefile target was added to regenerate the models: `make generate-python-api`

### Validation

go test ./...
Python import works: from model_spec.v1 import Model
Model generation works: make generate-python-api

fixes  #138

### CODE REVIEW

File: py/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+The models are generated from the canonical JSON Schema at `schema/config-schema.json` and are intended for downstream projects that need importable spec-aligned types.
+
+## Usage
+
+```python
+from model_spec.v1 import Model
+
+model = Model.model_validate_json(json_payload)
+print(model.descriptor.name)
+```
```
[the remark this answers is not in the document]
Comment (rishi-jat) [reply py/README.md:29]: done

**Comment by aftersnow** [comment]
Thanks for working on this. The schema-driven generation approach here looks useful, but it overlaps quite a bit with [#175](https://github.com/modelpack/model-spec/pull/175). I would suggest collaborating with #175 so we can converge on one Python SDK path and avoid maintaining two parallel Python APIs.

**Comment by rishi-jat** [comment]
After reviewing #175 and the current repo structure, here is my assessment.

The Python SDK in this PR is functional and well put together, but it introduces a second representation of the ModelPack specification by manually defining dataclasses that mirror the schema and Go types.

In this repository,` schema/config-schema.json` is already the canonical source of truth. The Go implementation aligns with that schema. Recreating the same structure manually in Python means the spec now exists in multiple places:
- JSON schema (canonical)
- Go structs
- Python dataclasses (this PR)

This creates a maintenance issue:
- Any schema change requires manual updates in Python
- There is no guarantee the Python layer stays aligned with the schema
- Drift between schema and SDK becomes likely over time

I also noticed:
- Schema loading is path-based, which is fragile for packaging/install use
- Validator correctness depends on strict schema dialect alignment
- Serialization logic is manually trying to mirror Go behavior, which is hard to maintain as the spec evolves

From what I see in the code and repo design, the direction looks schema-first rather than handwritten models.

An alternative that seems more aligned with the current design would be to keep the schema as the single source of truth and derive Python types from it, instead of maintaining handwritten spec types in parallel. Supporting pieces like validation, tests, and packaging can still be layered on top of that.

@aftersnow want to check your view on this:

Would it make more sense to avoid maintaining handwritten spec types here and instead keep the schema as the single source of truth, with Python types derived from it?

**Comment by pradhyum6144** [comment]
@rishi-jat these are valid points and I agree with most of them.

You're right that hand written dataclasses create a maintenance burden. I already acknowledged this in my earlier reply  I'm happy to adopt schema-driven generation as the base and layer my contributions on top.

Here's what I think a combined approach looks like:

From #184 (your PR): Schemadriven auto-generation via datamodel-code-generator  this becomes the foundation for Python types, keeping them in sync with config-schema.json automatically.

From #175 (my PR):

Validator with correct dialect alignment (Draft4Validator matching the schema's draft-04 declaration)
Test suite (64 tests covering serialization, validation, edge cases)
pyproject.toml packaging
CI workflow (PR #180  pytest across Python 3.10-3.13 + ruff linting)
I can rebase my PR to use your generated types as the base and keep the validator/tests/packaging on top. Or we can merge into one joint PR whatever works best for you and the maintainers.

**Comment by pradhyum6144** [comment]
@aftersnow what would you prefer two layered PRs or one combined effort???

**Comment by rishi-jat** [comment]
> @rishi-jat these are valid points and I agree with most of them.
> 
> You're right that hand written dataclasses create a maintenance burden. I already acknowledged this in my earlier reply I'm happy to adopt schema-driven generation as the base and layer my contributions on top.
> 
> Here's what I think a combined approach looks like:
> 
> From #184 (your PR): Schemadriven auto-generation via datamodel-code-generator this becomes the foundation for Python types, keeping them in sync with config-schema.json automatically.
> 
> From #175 (my PR):
> 
> Validator with correct dialect alignment (Draft4Validator matching the schema's draft-04 declaration) Test suite (64 tests covering serialization, validation, edge cases) pyproject.toml packaging CI workflow (PR #180 pytest across Python 3.10-3.13 + ruff linting) I can rebase my PR to use your generated types as the base and keep the validator/tests/packaging on top. Or we can merge into one joint PR whatever works best for you and the maintainers.

This direction makes sense to me.

This PR keeps the schema as the source of truth and generates Python models from it. The other PR’s validator, tests, and packaging can be layered on top instead of maintaining separate handwritten types.

@aftersnow does this approach look right?

**Review by bergwolf — COMMENTED** [review]
please add a CI job to ensure the generated python data structures are always updated, e.g., the CI job can regenerate them and compare with the existing one. If they differ, CI should fail.

**Comment by rishi-jat** [comment]
> please add a CI job to ensure the generated python data structures are always updated, e.g., the CI job can regenerate them and compare with the existing one. If they differ, CI should fail.

will do. Thanks!

**Comment by rishi-jat** [comment]
> please add a CI job to ensure the generated python data structures are always updated, e.g., the CI job can regenerate them and compare with the existing one. If they differ, CI should fail.

Added a CI job that regenerates the Python models from the schema and fails if any diff is detected. This ensures generated artifacts stay in sync with the schema and prevents drift.

**Comment by aagumin** [comment]
Any updates?

**Review by aftersnow — COMMENTED** [review]
Thanks for adding the regeneration check @bergwolf asked for. I ran the generator and the new CI steps locally — regeneration is byte-identical and the types map the schema correctly, so the pipeline itself is solid.

## Must fix

**1. The check can pass while the package is broken.** `__init__.py` is a hand-written re-export list, not generated and not import-tested. I renamed the `Modality` `$def`, regenerated, committed — `git diff --exit-code` returned 0 (CI green) but `from model_spec.v1 import Model` raised `ImportError`. Adding a new `$def` fails in reverse: CI passes, public API silently lags the spec. Fix: add a `python -c "from model_spec.v1 import Model"` step, or better a pytest running `docs/config.md` examples through the models (also covers #2).

**2. `model_dump_json()` emits JSON our schema rejects.** Unset optionals serialize as `null`, but the schema declares them `"type": "string"` — `jsonschema.validate` fails with `None is not of type 'string'`. Go uses `omitempty` so this is Python-only, and it breaks read/modify/write round-trips. Needs `exclude_none=True`, ideally baked in via `--base-class`.

## Should fix

- **No `pyproject.toml` or `py.typed`** — downstream still has to vendor, which is what #138 wanted to stop.
- **`RootModel` wrapping:** `inputTypes[0] == "text"` is `False`, callers need `.root`. Consider `--collapse-root-models`, or document it.
- **Versions pinned only in CI.** README/Makefile say `pip install datamodel-code-generator` unpinned; `black`/`isort` aren't pinned at all though they determine the output. A `py/requirements-dev.txt` used by all three would fix it.
- **`actions/checkout` mismatch:** new job v6.0.2, `lint` job above it v7.0.1.
- Missing Apache license headers (every `.go` file has one); `.gitignore` needs `__pycache__/`, `*.py[cod]`, `.venv/`.

## One warning

Don't take the `--snake-case-field` suggestion as-is — it emits aliases without `populate_by_name`, so `ModelDescriptor(doc_url=...)` fails to construct and dumps produce `diff_ids`, violating `additionalProperties`. It'd need `--allow-population-by-field-name` plus `by_alias=True`. I'd keep the schema field names.
